Update `GetPinnableReference()` usage in multi-step custom type marshalling

jkoritzinsky opened the pull request:
Change `GetPinnableReference()` on a marshaller type to be used as a side effect when marshalling in all cases when a `fixed()` statement is usable.

Use the `Value` property getter to get the value to pass to native in all cases.

Update our marshallers and tests to follow this design update. Also do some code cleanup along the way.

Fixes https://github.com/dotnet/runtimelab/issues/1653 in two ways:

1. Enables writing marshallers in C# source that can effectively solve that issue now that `GetPinnableReference` can return a value that needs to be pinned that is not equal to the value that is passed to native code.
2. Directly fixes that issue by preserving the statements within a `fixed()` statement returned in the Pin stage (was required to implement 1.).

jkoritzinsky committed:
Update GetPinnableReference() usage in multi-step custom tyep marshalling

Change GetPinnableReference() on a marshaller type to be used as a side effect when marshalling in all cases when a fixed() statement is usable.

Use the Value property getter to get the value to pass to native in all cases.

Update our marshallers and tests to follow this design update.

AaronRobinsonMSFT commented on docs/design/libraries/DllImportGenerator/StructMarshalling.md:
````diff
@@ -54,14 +54,16 @@ partial struct TNative
 ```
 
-The analyzer will report an error if neither the construtor nor the ToManaged method is defined. When one of those two methods is missing, the direction of marshalling (managed to native/native to managed) that relies on the missing method is considered unsupported for the corresponding managed type. The FreeNative method is only required when there are resources that need to be released.
+The analyzer will report an error if neither the constructor nor the ToManaged method is defined. When one of those two methods is missing, the direction of marshalling (managed to native/native to managed) that relies on the missing method is considered unsupported for the corresponding managed type. The FreeNative method is only required when there are resources that need to be released.
````
```suggestion
The analyzer will report an error if neither the constructor nor the `ToManaged` method is defined. When one of those two methods is missing, the direction of marshalling (managed to native/native to managed) that relies on the missing method is considered unsupported for the corresponding managed type. The `FreeNative` method is only required when there are resources that need to be released.
```

AaronRobinsonMSFT commented on docs/design/libraries/DllImportGenerator/StructMarshalling.md:
```diff
@@ -54,14 +54,16 @@ partial struct TNative
 If the native type `TNative` also has a public `Value` property, then the value of the `Value` property will be passed to native code instead of the `TNative` value itself. As a result, the type `TNative` will be allowed to be non-blittable and the type of the `Value` property will be required to be blittable. If the `Value` property is settable, then when marshalling in the native-to-managed dir ...
 
-A `ref` or `ref readonly` typed `Value` property is unsupported. If a ref-return is required, the type author can supply a `GetPinnableReference` method on the native type. If a `GetPinnableReference` method is supplied, then the `Value` property must have a pointer-sized primitive type.
+If a `Value` property is provided, the developer may also provide a ref-returning or readonly-ref-returning `GetPinnableReference` method. The `GetPinnableReference` method will be called before the `Value` property getter is called. The ref returned by `GetPinnableReference` will be pinned with a `fixed` statement, but the pinned will not be used (it acts exclusively as a side-effect).
```
```suggestion
If a `Value` property is provided, the developer may also provide a ref-returning or readonly-ref-returning `GetPinnableReference` method. The `GetPinnableReference` method will be called before the `Value` property getter is called. The ref returned by `GetPinnableReference` will be pinned with a `fixed` statement, but the pinned value will not be used (it acts exclusively as a side-effect).
```

AaronRobinsonMSFT commented on src/libraries/System.Runtime.InteropServices/gen/DllImportGenerator/AnalyzerReleases.Unshipped.md:
```diff
@@ -22,3 +22,4 @@ DLLIMPORTGENANALYZER014 | Usage            | Error    | RefValuePropertyUnsuppor
 DLLIMPORTGENANALYZER015 | Interoperability | Disabled | ConvertToGeneratedDllImportAnalyzer
 DLLIMPORTGENANALYZER016 | Usage            | Error    | GenericTypeMustBeClosed
 DLLIMPORTGENANALYZER017 | Usage            | Warning  | GeneratedDllImportContainingTypeMissingModifiers
+DLLIMPORTGENANALYZER018 | Usage            | Error  | MarshallerGetPinnableReferenceRequiresValueProperty
```
```suggestion
DLLIMPORTGENANALYZER018 | Usage            | Error    | MarshallerGetPinnableReferenceRequiresValueProperty
```

AaronRobinsonMSFT commented on src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/ICustomNativeTypeMarshallingStrategy.cs:
```diff
@@ -503,35 +500,34 @@ public IEnumerable<StatementSyntax> GenerateMarshalStatements(TypePositionInfo i
                 _valuePropertyType,
                 SingletonSeparatedList(
-                    VariableDeclarator(context.GetIdentifiers(info).native)
+                    VariableDeclarator(Identifier(context.GetAdditionalIdentifier(info, "ignored")))
```
Instead of `"ignored"` do we want to use `_`? I assume the later is the canonical way to do this in C#. Is there a reason we might want to be explicit here?

jkoritzinsky replied:
We can't use `_` as it's still a valid identifier name. As a result, if we have two parameters that both hit this path, we get errors about `_` having already been declared. 😢

AaronRobinsonMSFT replied:
boo.

jkoritzinsky committed:
Apply suggestions from code review

Co-authored-by: Aaron Robinson <arobins@microsoft.com>

jkoritzinsky committed:
Don't emit the assingment to the Value property in the Unmarshal stage when using [Out]

elinor-fung commented:
Do you have a good before/after example for the generated code?

jkoritzinsky commented:
Before:

```csharp
namespace DllImportGenerator.IntegrationTests
{
    partial class NativeExportsNE
    {
        public unsafe partial class Arrays
        {
            [System.Runtime.CompilerServices.SkipLocalsInitAttribute]
            public static partial int SumInArray(in int[] values, int numValues)
            {
                int __retVal;
                //
                // Setup
                //
                global::System.Runtime.InteropServices.GeneratedMarshalling.ArrayMarshaller<int> __values_gen_native__marshaller = default;
                try
                {
                    //
                    // Marshal
                    //
                    byte* __values_gen_native__marshaller__stackptr = stackalloc byte[global::System.Runtime.InteropServices.GeneratedMarshalling.ArrayMarshaller<int>.StackBufferSize];
                    __values_gen_native__marshaller = new(values, new System.Span<byte>(__values_gen_native__marshaller__stackptr, global::System.Runtime.InteropServices.GeneratedMarshalling.ArrayMarshaller<int>.StackBufferSize), sizeof(int));
                    __values_gen_native__marshaller.ManagedValues.CopyTo(System.Runtime.InteropServices.MemoryMarshal.Cast<byte, int>(__values_gen_native__marshaller.NativeValueStorage));
                    fixed (byte* __values_gen_native = &__values_gen_native__marshaller.GetPinnableReference())
                        __retVal = __PInvoke__(&__values_gen_native, numValues);
                }
                finally
                {
                    //
                    // Cleanup
                    //
                    __values_gen_native__marshaller.FreeNative();
                }

                return __retVal;
                //
                // Local P/Invoke
                //
                [System.Runtime.InteropServices.DllImportAttribute("Microsoft.Interop.Tests.NativeExportsNE", EntryPoint = "sum_int_array_ref")]
                extern static unsafe int __PInvoke__(byte** values, int numValues);
            }
        }
    }
}
```

After:

```
namespace DllImportGenerator.IntegrationTests
{
    partial class NativeExportsNE
    {
        public unsafe partial class Arrays
        {
            [System.Runtime.CompilerServices.SkipLocalsInitAttribute]
            public static partial int SumInArray(in int[] values, int numValues)
            {
                byte* __values_gen_native = default;
                int __retVal;
                //
                // Setup
                //
                global::System.Runtime.InteropServices.GeneratedMarshalling.ArrayMarshaller<int> __values_gen_native__marshaller = default;
                try
                {
                    //
                    // Marshal
                    //
                    byte* __values_gen_native__marshaller__stackptr = stackalloc byte[global::System.Runtime.InteropServices.GeneratedMarshalling.ArrayMarshaller<int>.StackBufferSize];
                    __values_gen_native__marshaller = new(values, new System.Span<byte>(__values_gen_native__marshaller__stackptr, global::System.Runtime.InteropServices.GeneratedMarshalling.ArrayMarshaller<int>.StackBufferSize), sizeof(int));
                    __values_gen_native__marshaller.ManagedValues.CopyTo(System.Runtime.InteropServices.MemoryMarshal.Cast<byte, int>(__values_gen_native__marshaller.NativeValueStorage));
                    fixed (byte* __values_gen_native__ignored = __values_gen_native__marshaller)
                    {
                        __values_gen_native = __values_gen_native__marshaller.Value;
                        {
                            __retVal = __PInvoke__(&__values_gen_native, numValues);
                        }
                    }
                }
                finally
                {
                    //
                    // Cleanup
                    //
                    __values_gen_native__marshaller.FreeNative();
                }

                return __retVal;
                //
                // Local P/Invoke
                //
                [System.Runtime.InteropServices.DllImportAttribute("Microsoft.Interop.Tests.NativeExportsNE", EntryPoint = "sum_int_array_ref")]
                extern static unsafe int __PInvoke__(byte** values, int numValues);
            }
        }
    }
}
```

jkoritzinsky committed:
Fix the unmarshalling conditions

jkoritzinsky committed:
Fix Utf16StringMarshaller to correctly handle the "null pointer" case now that we're using spans internally for storage in all cases.

jkoritzinsky committed:
Merge branch 'main' of github.com:dotnet/runtime into getpinnablereference-struct-marshalling-design-update
